File: cli/internal/analytics/analytics.go
// Package analytics turns raw transactions into the series the dashboard
// renders. Pure functions, no I/O (mirrors Analytics.swift).
package analytics

import (
	"sort"
	"time"

	"github.com/RomneyDa/phinny/cli/internal/model"
)

// MonthlyFlow is income vs spending for one calendar month.
type MonthlyFlow struct {
	Month   int64   `json:"month"` // epoch seconds of the first day of the month
	Income  float64 `json:"income"`
	Expense float64 `json:"expense"`
	Net     float64 `json:"net"`
}

// CategorySpend is spending grouped by category/merchant.
type CategorySpend struct {
	Label  string  `json:"label"`
	Amount float64 `json:"amount"`
}

// Summary holds the headline numbers for the summary cards.
type Summary struct {
	TotalBalance        float64 `json:"total_balance"`
	CurrentMonthIncome  float64 `json:"current_month_income"`
	CurrentMonthExpense float64 `json:"current_month_expense"`
	CurrentMonthNet     float64 `json:"current_month_net"`
	TransactionCount    int     `json:"transaction_count"`
	AccountCount        int     `json:"account_count"`
}

func startOfMonth(t time.Time) time.Time {
	y, m, _ := t.Date()
	return time.Date(y, m, 1, 0, 0, 0, 0, t.Location())
}

// MonthlyFlows returns income/expense per month for the last `months` months
// (oldest first), including empty months so the chart axis is continuous.
func MonthlyFlows(txns []model.Transaction, months int, now time.Time) []MonthlyFlow {
	if months <= 0 {
		months = 12
	}
	thisMonth := startOfMonth(now)

	buckets := make([]time.Time, 0, months)
	for offset := months - 1; offset >= 0; offset-- {
		buckets = append(buckets, thisMonth.AddDate(0, -offset, 0))
	}
	earliest := buckets[0]

	income := map[int64]float64{}
	expense := map[int64]float64{}
	for _, t := range txns {
		d := t.Date()
		if d.Before(earliest) {
			continue
		}
		key := startOfMonth(d).Unix()
		if t.Amount >= 0 {
			income[key] += t.Amount
		} else {
			expense[key] += -t.Amount
		}
	}

	out := make([]MonthlyFlow, 0, len(buckets))
	for _, b := range buckets {
		k := b.Unix()
		inc, exp := income[k], expense[k]
		out = append(out, MonthlyFlow{Month: k, Income: inc, Expense: exp, Net: inc - exp})
	}
	return out
}

// TopSpending returns the top spending groups within the last `days` days.
// Anything past topN is collapsed into "Other". `label` maps a transaction to
// its group label (e.g. effective category name).
func TopSpending(txns []model.Transaction, days, topN int, now time.Time, label func(model.Transaction) string) []CategorySpend {
	if days <= 0 {
		days = 30
	}
	if topN <= 0 {
		topN = 7
	}
	cutoff := now.AddDate(0, 0, -days)
	totals := map[string]float64{}
	for _, t := range txns {
		if !t.IsExpense() || t.Date().Before(cutoff) {
			continue
		}
		totals[label(t)] += -t.Amount
	}
	sorted := make([]CategorySpend, 0, len(totals))
	for k, v := range totals {
		sorted = append(sorted, CategorySpend{Label: k, Amount: v})
	}
	sort.Slice(sorted, func(i, j int) bool {
		if sorted[i].Amount != sorted[j].Amount {
			return sorted[i].Amount > sorted[j].Amount
		}
		return sorted[i].Label < sorted[j].Label
	})
	if len(sorted) <= topN {
		return sorted
	}
	top := append([]CategorySpend(nil), sorted[:topN]...)
	other := 0.0
	for _, s := range sorted[topN:] {
		other += s.Amount
	}
	return append(top, CategorySpend{Label: "Other", Amount: other})
}

// SummaryOf computes the headline numbers.
func SummaryOf(accounts []model.Account, txns []model.Transaction, now time.Time) Summary {
	s := Summary{AccountCount: len(accounts), TransactionCount: len(txns)}
	for _, a := range accounts {
		s.TotalBalance += a.Balance
	}
	monthStart := startOfMonth(now)
	for _, t := range txns {
		if t.Date().Before(monthStart) {
			continue
		}
		if t.Amount >= 0 {
			s.CurrentMonthIncome += t.Amount
		} else {
			s.CurrentMonthExpense += -t.Amount
		}
	}
	s.CurrentMonthNet = s.CurrentMonthIncome - s.CurrentMonthExpense
	return s
}

File: archive/Sources-swift-engine/Analytics.swift
import Foundation

/// Pure aggregation helpers that turn raw transactions into the series the
/// charts render. No I/O - easy to read, easy to test.
enum Analytics {

    /// Income vs. spending for one calendar month.
    struct MonthlyFlow: Identifiable {
        let month: Date            // first day of the month
        let income: Double         // sum of positive amounts
        let expense: Double        // sum of |negative amounts|
        var id: Date { month }
        var net: Double { income - expense }
    }

    /// Spending grouped by category/merchant.
    struct CategorySpend: Identifiable {
        let label: String
        let amount: Double
        var id: String { label }
    }

    /// Headline numbers for the summary cards.
    struct Summary {
        var totalBalance: Double = 0
        var currentMonthIncome: Double = 0
        var currentMonthExpense: Double = 0
        var transactionCount: Int = 0
        var accountCount: Int = 0
        var currentMonthNet: Double { currentMonthIncome - currentMonthExpense }
    }

    private static var calendar: Calendar {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = .current
        return c
    }

    private static func startOfMonth(_ date: Date) -> Date {
        let c = calendar
        return c.date(from: c.dateComponents([.year, .month], from: date)) ?? date
    }

    /// Income/expense per month for the last `months` months (oldest first),
    /// including empty months so the chart has a continuous axis.
    static func monthlyFlows(_ transactions: [Transaction], months: Int = 12) -> [MonthlyFlow] {
        let cal = calendar
        let thisMonth = startOfMonth(Date())

        // Build the ordered list of buckets we want to show.
        var buckets: [Date] = []
        for offset in stride(from: months - 1, through: 0, by: -1) {
            if let m = cal.date(byAdding: .month, value: -offset, to: thisMonth) {
                buckets.append(m)
            }
        }
        let earliest = buckets.first ?? thisMonth

        var income: [Date: Double] = [:]
        var expense: [Date: Double] = [:]
        for txn in transactions where txn.date >= earliest {
            let key = startOfMonth(txn.date)
            if txn.amount >= 0 { income[key, default: 0] += txn.amount }
            else { expense[key, default: 0] += -txn.amount }
        }

        return buckets.map { month in
            MonthlyFlow(month: month, income: income[month] ?? 0, expense: expense[month] ?? 0)
        }
    }

    /// Top spending groups (category → merchant → description) within the last
    /// `days` days. Anything past `topN` is collapsed into "Other".
    static func topSpending(_ transactions: [Transaction], days: Int = 30, topN: Int = 7,
                            label: (Transaction) -> String = { $0.groupLabel }) -> [CategorySpend] {
        let cutoff = calendar.date(byAdding: .day, value: -days, to: Date()) ?? .distantPast
        var totals: [String: Double] = [:]
        for txn in transactions where txn.isExpense && txn.date >= cutoff {
            totals[label(txn), default: 0] += -txn.amount
        }
        let sorted = totals.map { CategorySpend(label: $0.key, amount: $0.value) }
            .sorted { $0.amount > $1.amount }

        guard sorted.count > topN else { return sorted }
        let top = Array(sorted.prefix(topN))
        let otherTotal = sorted.dropFirst(topN).reduce(0) { $0 + $1.amount }
        return top + [CategorySpend(label: "Other", amount: otherTotal)]
    }

    static func summary(accounts: [Account], transactions: [Transaction]) -> Summary {
        var s = Summary()
        s.accountCount = accounts.count
        s.transactionCount = transactions.count
        s.totalBalance = accounts.reduce(0) { $0 + $1.balance }

        let monthStart = startOfMonth(Date())
        for txn in transactions where txn.date >= monthStart {
            if txn.amount >= 0 { s.currentMonthIncome += txn.amount }
            else { s.currentMonthExpense += -txn.amount }
        }
        return s
    }
}
